core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1103,9 +1103,14 @@ class DAGScheduler(
         // multiple tasks running concurrently on different executors). In that case, it is possible
         // the fetch failure has already been handled by the scheduler.
         if (runningStages.contains(failedStage)) {
-          logInfo(s"Marking $failedStage (${failedStage.name}) as failed " +
-            s"due to a fetch failure from $mapStage (${mapStage.name})")
-          markStageAsFinished(failedStage, Some(failureMessage))
+          if (failedStage.attemptId - 1 > task.stageAttemptId) {
```

Comment:
this is really the only important part of the fix -- if we get a fetch failure, but we've already failed the attempt of the stage that caused the fetch failure, then do *not* resubmit the stage again.  (The rest of the changes are just adding the stageAttemptId to the task)